Source/SwiftLintFramework/Extensions/Array+SwiftLint.swift

```diff
@@ -59,18 +59,16 @@ extension Array {
         return parallelMap(transform: transform).flatMap { $0 }
     }
 
-    func parallelMap<T>(transform: @escaping ((Element) -> T)) -> [T] {
-        var result = [(Int, T)]()
-        result.reserveCapacity(count)
-
-        let queueLabelPrefix = "io.realm.SwiftLintFramework.map.\(NSUUID().uuidString)"
-        let resultAccumulatorQueue = DispatchQueue(label: "\(queueLabelPrefix).resultAccumulator")
-        DispatchQueue.concurrentPerform(iterations: count) { index in
-            let jobIndexAndResults = (index, transform(self[index]))
+    func parallelMap<T>(transform: (Element) -> T) -> [T] {
+        var result = [T?](repeating: nil, count: count)
+        let resultAccumulatorQueue = DispatchQueue(label: "io.realm.SwiftLintFramework.map.resultAccumulator")
+        DispatchQueue.concurrentPerform(iterations: count) { idx in
+            let element = self[idx]
+            let transformed = transform(element)
             resultAccumulatorQueue.sync {
```

Comment:
Is this `sync` still necessary?

Reply:
I think so

Reply:
I understand that CoW will happen without `sync`.
`sync` free implementation will be following:
```swift
    func parallelMap<T>(transform: (Element) -> T) -> [T] {
        var result = ContiguousArray<T?>(repeating: nil, count: count)
        return result.withUnsafeMutableBufferPointer { buffer in
            DispatchQueue.concurrentPerform(iterations: buffer.count) { idx in
                buffer[idx] = transform(self[idx])
            }
            return buffer.map { $0! }
        }
    }
```

Reply:
Nice! And Thread Sanitizer is happy with that? If so, mind filing a PR?

Reply:
odd

Reply:
```terminal.sh-session
$ DEVELOPER_DIR=/Applications/Xcode.app xcodebuild -workspace SwiftLint.xcworkspace/ -scheme swiftlint -enableThreadSanitizer YES test -only-testing:SwiftLintFrameworkTests/IntegrationTests/testSwiftLintLints
…
Test Case '-[SwiftLintFrameworkTests.IntegrationTests testSwiftLintLints]' started.
Loading configuration from '.swiftlint.yml'
==================
WARNING: ThreadSanitizer: data race (pid=32964)
  Read of size 8 at 0x7b1800019310 by main thread:
  * #0 File.lines.getter File.swift:37 (SourceKittenFramework:x86_64+0x3c537)
    #1 TrailingWhitespaceRule.validate(file:) TrailingWhitespaceRule.swift:32 (SwiftLintFramework:x86_64+0x4be08c)
    #2 protocol witness for Rule.validate(file:) in conformance TrailingWhitespaceRule TrailingWhitespaceRule.swift (SwiftLintFramework:x86_64+0x4c2d80)
    #3 Rule.lint(file:regions:benchmark:superfluousDisableCommandRule:) Linter.swift:71 (SwiftLintFramework:x86_64+0xc1119)
    #4 closure #2 in Linter.getStyleViolations(benchmark:) Linter.swift:138 (SwiftLintFramework:x86_64+0xd3acc)
    #5 partial apply for closure #2 in Linter.getStyleViolations(benchmark:) Linter.swift (SwiftLintFramework:x86_64+0xdc5dc)
    #6 thunk for @callee_owned (@in Rule) -> (@owned LintResult?) Linter.swift (SwiftLintFramework:x86_64+0xd3d9a)
    #7 partial apply for thunk for @callee_owned (@in Rule) -> (@owned LintResult?) Linter.swift (SwiftLintFramework:x86_64+0xdc6c9)
    #8 closure #1 in Array.parallelMap<A>(transform:) Array+SwiftLint.swift:67 (SwiftLintFramework:x86_64+0x4a966e)
    #9 partial apply for closure #1 in Array.parallelMap<A>(transform:) Array+SwiftLint.swift (SwiftLintFramework:x86_64+0x4a99c4)
    #10 thunk for @callee_owned (@unowned Int) -> () <null> (libswiftDispatch.dylib:x86_64+0xb5a6)
    #11 _dispatch_client_callout2 <null> (libdispatch.dylib:x86_64+0xc167)
    #12 _swift_dispatch_apply_current <null> (libswiftDispatch.dylib:x86_64+0x12f51)
    #13 Array.parallelFlatMap<A>(transform:) Array+SwiftLint.swift:59 (SwiftLintFramework:x86_64+0x4a91b5)
    #14 Linter.getStyleViolations(benchmark:) Linter.swift:139 (SwiftLintFramework:x86_64+0xcfa8b)
    #15 Linter.styleViolations.getter Linter.swift:116 (SwiftLintFramework:x86_64+0xce74f)
    #16 closure #2 in IntegrationTests.testSwiftLintLints() IntegrationTests.swift:31 (SwiftLintFrameworkTests:x86_64+0x1692b)
    #17 thunk for @callee_owned (@owned File) -> (@owned [StyleViolation], @error @owned Error) IntegrationTests.swift (SwiftLintFrameworkTests:x86_64+0x16ae1)
    #18 partial apply for thunk for @callee_owned (@owned File) -> (@owned [StyleViolation], @error @owned Error) IntegrationTests.swift (SwiftLintFrameworkTests:x86_64+0x16c5a)
    #19 Sequence.flatMap<A>(_:) <null> (libswiftCore.dylib:x86_64+0x15f752)
    #20 @objc IntegrationTests.testSwiftLintLints() IntegrationTests.swift (SwiftLintFrameworkTests:x86_64+0x18930)
    #21 __invoking___ <null> (CoreFoundation:x86_64h+0x829db)

  Previous write of size 8 at 0x7b1800019310 by thread T4 (mutexes: write M2476):
  * #0 File.lines.getter File.swift:38 (SourceKittenFramework:x86_64+0x3c682)
    #1 File.syntaxTokensByLine() File+SwiftLint.swift:155 (SwiftLintFramework:x86_64+0x29f721)
    #2 closure #1 in variable initialization expression of syntaxTokensByLinesCache File+Cache.swift:31 (SwiftLintFramework:x86_64+0x27d869)
    #3 thunk for @callee_owned (@owned File) -> (@owned [[SyntaxToken]]?) File+Cache.swift (SwiftLintFramework:x86_64+0x27d8e3)
    #4 partial apply for thunk for @callee_owned (@owned File) -> (@owned [[SyntaxToken]]?) File+Cache.swift (SwiftLintFramework:x86_64+0x2834f9)
    #5 Cache.get(_:) File+Cache.swift:73 (SwiftLintFramework:x86_64+0x27f132)
    #6 File.syntaxTokensByLines.getter File+Cache.swift:152 (SwiftLintFramework:x86_64+0x281a48)
    #7 ClosureSpacingRule.validBraces(in:) ClosureSpacingRule.swift:84 (SwiftLintFramework:x86_64+0x447b72)
    #8 ClosureSpacingRule.findViolations(file:) ClosureSpacingRule.swift:129 (SwiftLintFramework:x86_64+0x449b09)
    #9 ClosureSpacingRule.validate(file:) ClosureSpacingRule.swift:146 (SwiftLintFramework:x86_64+0x44b91a)
    #10 protocol witness for Rule.validate(file:) in conformance ClosureSpacingRule ClosureSpacingRule.swift (SwiftLintFramework:x86_64+0x44f874)
    #11 Rule.lint(file:regions:benchmark:superfluousDisableCommandRule:) Linter.swift:71 (SwiftLintFramework:x86_64+0xc1119)
    #12 closure #2 in Linter.getStyleViolations(benchmark:) Linter.swift:138 (SwiftLintFramework:x86_64+0xd3acc)
    #13 partial apply for closure #2 in Linter.getStyleViolations(benchmark:) Linter.swift (SwiftLintFramework:x86_64+0xdc5dc)
    #14 thunk for @callee_owned (@in Rule) -> (@owned LintResult?) Linter.swift (SwiftLintFramework:x86_64+0xd3d9a)
    #15 partial apply for thunk for @callee_owned (@in Rule) -> (@owned LintResult?) Linter.swift (SwiftLintFramework:x86_64+0xdc6c9)
    #16 closure #1 in Array.parallelMap<A>(transform:) Array+SwiftLint.swift:67 (SwiftLintFramework:x86_64+0x4a966e)
    #17 partial apply for closure #1 in Array.parallelMap<A>(transform:) Array+SwiftLint.swift (SwiftLintFramework:x86_64+0x4a99c4)
    #18 thunk for @callee_owned (@unowned Int) -> () <null> (libswiftDispatch.dylib:x86_64+0xb5a6)
    #19 _dispatch_client_callout2 <null> (libdispatch.dylib:x86_64+0xc167)

  Issue is caused by frames marked with "*".

  Location is heap block of size 88 at 0x7b18000192c0 allocated by main thread:
    #0 malloc <null> (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x48b3a)
    #1 swift_slowAlloc <null> (libswiftCore.dylib:x86_64+0x344478)
    #2 File.__allocating_init(pathDeferringReading:) File.swift (SourceKittenFramework:x86_64+0x3d6cc)
    #3 partial apply for File.__allocating_init(pathDeferringReading:) Configuration+LintableFiles.swift (SwiftLintFramework:x86_64+0x2d303e)
    #4 thunk for @callee_owned (@owned String) -> (@owned File?, @error @owned Error) Configuration+LintableFiles.swift (SwiftLintFramework:x86_64+0x2d0d29)
    #5 partial apply for thunk for @callee_owned (@owned String) -> (@owned File?, @error @owned Error) Configuration+LintableFiles.swift (SwiftLintFramework:x86_64+0x2d0eaa)
    #6 Sequence.flatMap<A>(_:) <null> (libswiftCore.dylib:x86_64+0x15fa87)
    #7 IntegrationTests.testSwiftLintLints() IntegrationTests.swift:27 (SwiftLintFrameworkTests:x86_64+0x15514)
    #8 @objc IntegrationTests.testSwiftLintLints() IntegrationTests.swift (SwiftLintFrameworkTests:x86_64+0x18930)
    #9 __invoking___ <null> (CoreFoundation:x86_64h+0x829db)

  Mutex M2476 (0x7b18000066d0) created at:
    #0 pthread_mutex_init <null> (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x2b213)
    #1 -[NSLock init] <null> (Foundation:x86_64+0x1d7d)
    #2 NSLock.__allocating_init() Configuration+Cache.swift (SwiftLintFramework:x86_64+0x26d1f4)
    #3 Cache.init(_:) File+Cache.swift:60 (SwiftLintFramework:x86_64+0x27ead7)
    #4 Cache.__allocating_init(_:) File+Cache.swift (SwiftLintFramework:x86_64+0x27c0eb)
    #5 globalinit_33_D93545007D2CCCB97554C3EFBF0F3773_func4 File+Cache.swift:31 (SwiftLintFramework:x86_64+0x27d7f2)
    #6 dispatch_once <null> (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x66e04)
    #7 dispatch_once_f <null> (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x66ef6)
    #8 syntaxTokensByLinesCache.unsafeMutableAddressor File+Cache.swift:31 (SwiftLintFramework:x86_64+0x27d969)
    #9 File.syntaxTokensByLines.getter File+Cache.swift:152 (SwiftLintFramework:x86_64+0x2819e7)
    #10 ClosureSpacingRule.validBraces(in:) ClosureSpacingRule.swift:84 (SwiftLintFramework:x86_64+0x447b72)
    #11 ClosureSpacingRule.findViolations(file:) ClosureSpacingRule.swift:129 (SwiftLintFramework:x86_64+0x449b09)
    #12 ClosureSpacingRule.validate(file:) ClosureSpacingRule.swift:146 (SwiftLintFramework:x86_64+0x44b91a)
    #13 protocol witness for Rule.validate(file:) in conformance ClosureSpacingRule ClosureSpacingRule.swift (SwiftLintFramework:x86_64+0x44f874)
    #14 Rule.lint(file:regions:benchmark:superfluousDisableCommandRule:) Linter.swift:71 (SwiftLintFramework:x86_64+0xc1119)
    #15 closure #2 in Linter.getStyleViolations(benchmark:) Linter.swift:138 (SwiftLintFramework:x86_64+0xd3acc)
    #16 partial apply for closure #2 in Linter.getStyleViolations(benchmark:) Linter.swift (SwiftLintFramework:x86_64+0xdc5dc)
    #17 thunk for @callee_owned (@in Rule) -> (@owned LintResult?) Linter.swift (SwiftLintFramework:x86_64+0xd3d9a)
    #18 partial apply for thunk for @callee_owned (@in Rule) -> (@owned LintResult?) Linter.swift (SwiftLintFramework:x86_64+0xdc6c9)
    #19 closure #1 in Array.parallelMap<A>(transform:) Array+SwiftLint.swift:67 (SwiftLintFramework:x86_64+0x4a966e)
    #20 partial apply for closure #1 in Array.parallelMap<A>(transform:) Array+SwiftLint.swift (SwiftLintFramework:x86_64+0x4a99c4)
    #21 thunk for @callee_owned (@unowned Int) -> () <null> (libswiftDispatch.dylib:x86_64+0xb5a6)
    #22 _dispatch_client_callout2 <null> (libdispatch.dylib:x86_64+0xc167)

  Thread T4 (tid=710935, running) is a GCD worker thread

SUMMARY: ThreadSanitizer: data race File.swift:37 in File.lines.getter
…
```

Reply:
filed https://github.com/realm/SwiftLint/issues/2089

Reply:
filed a PR https://github.com/realm/SwiftLint/pull/2090